scheduler/README.md

```diff
@@ -0,0 +1,192 @@
+               │                expiring.
+               └───────┐        - filterByRescheduleable: updates
+                       │        the untainted bucket and creates
+                       │        2 new ones: rescheduleNow and
+                       │        rescheduleLater
+                       │        - reconcileReconnecting: returns
+                       │        which allocs should be marked for
+                       │        reconnecting and which should be
+                       │        stopped
+                       │        - computeStop
+                       │        - computeCanaries
+                       │        - computePlacements: allocs are
+                       │        placed if deployment is not
+                       │        paused or failed, they are not
+                       │        canaries (unless promoted),
+                       │        previous alloc was lost
+                       │        - computeReplacements
+                       │        - createDeployment
+                       │
+                       ▼
+┌────────────────────────────────────────────┐
+│computeDeploymentUpdates(deploymentComplete)│
+└────────────────────────────────────────────┘
+                       │
+                ┌──────┘        for complete deployments, it
+                │               handles multi-region case and
+                ▼               sets the deploymentUpdates
+        ┌───────────────┐
+        │return a.result│
+        └───────────────┘
+```
+
+## Feasibility checking
+
+Nomad uses a set of iterators to iterate over all nodes and check how feasible
+they are for any given job. This code sits in `scheduler/feasible.go` and
+`scheduler/stack.go` files. There are two interfaces: `FeasibleIterator` and
+`FeasibilityChecker`, and various objects that implement iterating over nodes to
+match constraints.
+
+Feasibility checkers are registered in the `Stack`, an interface implemented by
+`GenericStack` and `SystemStack`, for service and batch jobs, and for system and
+sysbatch jobs, respectively.
+
+## Node reconciliation
+
+The system scheduler also does a "reconciliation" step, but only on a
+per-node basis (system jobs run on all feasible nodes), which makes it
+simpler than the service reconciler which takes into account a whole cluster,
+and has jobs that can run on arbitrary subset of clients. The code is in
+`scheduler/scheduler_system.go` file.
+
+Node reconciliation removes tainted nodes, updates terminal allocations to lost,
+deals with disconnected nodes and computes placements.
+
+## Finding the best fit and scoring
+
+Applies only to service and batch jobs, since system and sysbatch jobs are
+placed on all feasible nodes.
+
+This part of scheduling sits in the `scheduler/rank.go` file. The `RankIterator`
+interface, which is implemented by e.g., `SpreadIterator` and `BinPackIterator`,
+captures the ranking logic in its `Next()` methods.
```

Comment:
I suspect a more detailed explanation of how fitting and scoring work is in order, but it could become a separate PR?